cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -0,0 +1,2987 @@
+                sink.put1(mkModRegRM(0, encG & 7, 4));
+                sink.put1(0x24);
+            } else if low8willSXto32(*simm32) && encE != regs::ENC_RSP && encE != regs::ENC_R12 {
+                sink.put1(mkModRegRM(1, encG & 7, encE & 7));
+                sink.put1((simm32 & 0xFF) as u8);
+            } else if encE != regs::ENC_RSP && encE != regs::ENC_R12 {
+                sink.put1(mkModRegRM(2, encG & 7, encE & 7));
+                sink.put4(*simm32);
+            } else if (encE == regs::ENC_RSP || encE == regs::ENC_R12) && low8willSXto32(*simm32) {
+                // REX.B distinguishes RSP from R12
+                sink.put1(mkModRegRM(1, encG & 7, 4));
+                sink.put1(0x24);
+                sink.put1((simm32 & 0xFF) as u8);
+            } else if encE == regs::ENC_R12 || encE == regs::ENC_RSP {
+                //.. wait for test case for RSP case
+                // REX.B distinguishes RSP from R12
+                sink.put1(mkModRegRM(2, encG & 7, 4));
+                sink.put1(0x24);
+                sink.put4(*simm32);
+            } else {
+                panic!("emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE: IR");
+            }
+        }
+        // Bizarrely, the IRRS case is much simpler.
+        Addr::IRRS {
+            simm32,
+            base: regBase,
+            index: regIndex,
+            shift,
+        } => {
+            let encBase = iregEnc(*regBase);
+            let encIndex = iregEnc(*regIndex);
+            // The rex byte
+            let w = if clearRexW { 0 } else { 1 };
+            let r = (encG >> 3) & 1;
+            let x = (encIndex >> 3) & 1;
+            let b = (encBase >> 3) & 1;
+            let rex = 0x40 | (w << 3) | (r << 2) | (x << 1) | b;
+            if rex != 0x40 || retainRedundant {
+                sink.put1(rex);
+            }
+            // All other prefixes and opcodes
+            while numOpcodes > 0 {
+                numOpcodes -= 1;
+                sink.put1(((opcodes >> (numOpcodes << 3)) & 0xFF) as u8);
+            }
+            // modrm, SIB, immediates
+            if low8willSXto32(*simm32) && encIndex != regs::ENC_RSP {
+                sink.put1(mkModRegRM(1, encG & 7, 4));
+                sink.put1(mkSIB(*shift, encIndex & 7, encBase & 7));
+                sink.put1(*simm32 as u8);
+            } else if encIndex != regs::ENC_RSP {
+                sink.put1(mkModRegRM(2, encG & 7, 4));
+                sink.put1(mkSIB(*shift, encIndex & 7, encBase & 7));
+                sink.put4(*simm32);
+            } else {
+                panic!("emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE: IRRS");
+            }
+        }
+    }
+}
+
+// This is the core 'emit' function for instructions that do not reference
+// memory.
+//
+// This is conceptually the same as
+// emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE, except it is for the case
+// where the E operand is a register rather than memory.  Hence it is much
+// simpler.
```

Comment:
```suggestion
/// This is the core 'emit' function for instructions that do not reference
/// memory.
///
/// This is conceptually the same as
/// emit_REX_OPCODES_MODRM_SIB_IMM_encG_memE, except it is for the case
/// where the E operand is a register rather than memory.  Hence it is much
/// simpler.
```